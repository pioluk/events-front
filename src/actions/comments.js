// @flow

import {
  FETCH_COMMENTS,
  FETCH_COMMENTS_SUCCESS,
  FETCH_COMMENTS_FAILURE,
  ADD_COMMENT,
  ADD_COMMENT_SUCCESS,
  ADD_COMMENT_FAILURE,
  REMOVE_COMMENT,
  REMOVE_COMMENT_SUCCESS,
  REMOVE_COMMENT_FAILURE,
} from '../constants/comments'

export function fetchComments (eventId: number, limit: number, offset: number) {
  return {
    type: FETCH_COMMENTS,
    eventId,
    limit,
    offset
  }
}

export function fetchCommentsSuccess (count: number, comments: any[]) {
  return {
    type: FETCH_COMMENTS_SUCCESS,
    count,
    comments
  }
}

export function fetchCommentsFailure (error: Error) {
  return {
    type: FETCH_COMMENTS_FAILURE,
    error
  }
}

export function addComment (text: string) {
  return {
    type: ADD_COMMENT,
    text
  }
}

export function addCommentSuccess (comment: any) {
  return {
    type: ADD_COMMENT_SUCCESS,
    comment
  }
}

export function addCommentFailure (error: Error) {
  return {
    type: ADD_COMMENT_FAILURE,
    error
  }
}

export function removeComment (id: number) {
  return {
    type: REMOVE_COMMENT,
    id
  }
}

export function removeCommentSuccess (id: number) {
  return {
    type: REMOVE_COMMENT_SUCCESS,
    id
  }
}

export function removeCommentFailure (error: Error) {
  return {
    type: REMOVE_COMMENT_FAILURE,
    error
  }
}
