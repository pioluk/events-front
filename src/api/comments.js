import { apiGet, apiPost, apiDelete } from './utils'

export const getComments = (eventId: number) =>
  apiGet(`event/${eventId}/comment`)

export const createComment = (eventId: number, text: string) =>
  apiPost(`event/${eventId}/comment`, { text })

export const removeComment = (eventId: number, commentId: number) =>
  apiDelete(`event/${eventId}/comment/${commentId}`)
