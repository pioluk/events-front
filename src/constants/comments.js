export const FETCH_COMMENTS = 'FETCH_COMMENTS'
export const FETCH_COMMENTS_SUCCESS = 'FETCH_COMMENTS_SUCCESS'
export const FETCH_COMMENTS_FAILURE = 'FETCH_COMMENTS_FAILURE'
export const ADD_COMMENT = 'ADD_COMMENT'
export const ADD_COMMENT_SUCCESS = 'ADD_COMMENT_SUCCESS'
export const ADD_COMMENT_FAILURE = 'ADD_COMMENT_FAILURE'
export const REMOVE_COMMENT = 'REMOVE_COMMENT'
export const REMOVE_COMMENT_SUCCESS = 'REMOVE_COMMENT_SUCCESS'
export const REMOVE_COMMENT_FAILURE = 'REMOVE_COMMENT_FAILURE'
export const INCREASE_PAGE = 'INCREASE_PAGE'
export const RESET_COMMENTS = 'RESET_COMMENTS'
