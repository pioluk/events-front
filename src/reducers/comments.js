// @flow

import {
  FETCH_COMMENTS,
  FETCH_COMMENTS_SUCCESS,
  FETCH_COMMENTS_FAILURE,
  ADD_COMMENT,
  ADD_COMMENT_SUCCESS,
  ADD_COMMENT_FAILURE,
  REMOVE_COMMENT,
  REMOVE_COMMENT_SUCCESS,
  REMOVE_COMMENT_FAILURE
} from '../constants/comments'

type Comment = any

type CommentsState = {
  count: number,
  comments: Array<Comment>,
  error: ?Error,
  isLoading: boolean
}

const initialState: CommentsState = {
  count: 0,
  comments: [],
  error: null,
  isLoading: false
}

export default function commentsReducer (state: CommentsState = initialState, action: any) {
  switch (action.type) {
    case FETCH_COMMENTS:
      return { ...state, isLoading: true }

    case FETCH_COMMENTS_SUCCESS:
      return { ...state, count: action.count, comments: action.comments, isLoading: false }

    case FETCH_COMMENTS_FAILURE:
      return { ...state, error: action.error, isLoading: false }

    case ADD_COMMENT:
      return { ...state }

    case ADD_COMMENT_SUCCESS:
      return {
        ...state,
        comments: [action.comment, ...state.comments]
      }

    case ADD_COMMENT_FAILURE:
      return { ...state, error: action.error }

    case REMOVE_COMMENT_SUCCESS: {
      const comments = state.comments.filter(comment => comment.id !== action.id)
      return { ...state, comments }
    }

    case REMOVE_COMMENT_FAILURE:
      return { ...state, error: action.error }

    default:
      return state
  }
}
