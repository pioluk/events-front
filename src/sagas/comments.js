import { call, take, put } from 'redux-saga/effects'
import { FETCH_COMMENTS, ADD_COMMENT, REMOVE_COMMENT } from '../constants/comments'
import * as actions from '../actions/comments'
import { getComments, createComment, removeComment } from '../api/comments'

export function* getCommentsFlow () {
  while (true) {
    const { eventId, limit, offset } = yield take(FETCH_COMMENTS)
    try {
      const { count, comments } = yield call(getComments, eventId, limit, offset)
      if (comments) {
        yield put(actions.fetchCommentsSuccess(count, comments))
      }
    } catch (err) {
      yield put(actions.fetchCommentsFailure(err))
    }
  }
}

export function* createCommentFlow () {
  while (true) {
    const { eventId, text } = yield take(ADD_COMMENT)
    try {
      const comment = yield call(createComment, eventId, text)
      if (comment) {
        yield put(actions.addCommentSuccess(comment))
      }
    } catch (err) {
      yield put(actions.addCommentFailure(err))
    }
  }
}

export function* removeCommentFlow () {
  while (true) {
    const { id } = yield take(REMOVE_COMMENT)
    try {
      yield call(removeComment, id)
      yield put(actions.removeCommentSuccess(id))
    } catch (err) {
      yield put(actions.removeCommentFailure(err))
    }
  }
}
