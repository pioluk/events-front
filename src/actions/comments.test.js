// flow

import * as actions from './comments'

describe('comments actions', () => {
  it('should create an action to fetch comments', () => {
    expect(actions.fetchComments(1, 20, 0)).toMatchSnapshot()
  })

  it('should create an action to send fetched comments', () => {
    const comments = [
      {
        user: {
          id: 1,
          username: 'pioluk',
          email: '181423@edu.p.lodz.pl',
          imageAvatar: null,
          createdAt: '2016-12-01T13:02:38.213Z',
          updatedAt: '2016-12-01T13:02:38.213Z'
        },
        id: 1,
        UserId: 1,
        EventId: 1,
        text: 'Lorem ipsum',
        createdAt: '2016-12-03T12:00:00.000Z',
        updatedAt: '2016-12-03T12:00:00.000Z',
        deletedAt: null
      },
      {
        user: {
          id: 1,
          username: 'pioluk',
          email: '181423@edu.p.lodz.pl',
          imageAvatar: null,
          createdAt: '2016-12-01T13:02:38.213Z',
          updatedAt: '2016-12-01T13:02:38.213Z'
        },
        id: 2,
        UserId: 1,
        EventId: 1,
        text: 'Dolor Amet',
        createdAt: '2016-12-03T12:30:00.000Z',
        updatedAt: '2016-12-03T12:30:00.000Z',
        deletedAt: null
      }
    ]
    expect(actions.fetchCommentsSuccess(2, comments)).toMatchSnapshot()
  })

  it('should create an action to indicate comment fetching error', () => {
    const error = new Error('Error fetching comments')
    expect(actions.fetchCommentsFailure(error)).toMatchSnapshot()
  })

  it('should create an action representing text to create comment', () => {
    const eventId = 1
    const text = 'Comment #1'
    expect(actions.addComment(eventId, text)).toMatchSnapshot()
  })

  it('should create an action representing a comment to add', () => {
    const comment = {
      user: {
        id: 1,
        username: 'pioluk',
        email: '181423@edu.p.lodz.pl',
        imageAvatar: null,
        createdAt: '2016-12-01T13:02:38.213Z',
        updatedAt: '2016-12-01T13:02:38.213Z'
      },
      id: 1,
      UserId: 1,
      EventId: 1,
      text: 'Lorem ipsum',
      createdAt: '2016-12-03T12:00:00.000Z',
      updatedAt: '2016-12-03T12:00:00.000Z',
      deletedAt: null
    }
    expect(actions.addCommentSuccess(comment)).toMatchSnapshot()
  })

  it('should create an action representing comment adding error', () => {
    const error = new Error('Comment adding failed')
    expect(actions.addCommentFailure(error)).toMatchSnapshot()
  })

  it('should create an action representing a request to remove a comment', () => {
    const id = 1
    expect(actions.removeComment(id)).toMatchSnapshot()
  })

  it('should create an action representing a comment to remove', () => {
    const id = 1
    expect(actions.removeCommentSuccess(id)).toMatchSnapshot()
  })

  it('should create an action representing comment removal erorr', () => {
    const error = new Error('Comment removing failed')
    expect(actions.removeCommentFailure(error)).toMatchSnapshot()
  })
})
