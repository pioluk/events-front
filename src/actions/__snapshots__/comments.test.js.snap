exports[`comments actions should create an action representing a comment to add 1`] = `
Object {
  "comment": Object {
    "EventId": 1,
    "UserId": 1,
    "createdAt": "2016-12-03T12:00:00.000Z",
    "deletedAt": null,
    "id": 1,
    "text": "Lorem ipsum",
    "updatedAt": "2016-12-03T12:00:00.000Z",
    "user": Object {
      "createdAt": "2016-12-01T13:02:38.213Z",
      "email": "181423@edu.p.lodz.pl",
      "id": 1,
      "imageAvatar": null,
      "updatedAt": "2016-12-01T13:02:38.213Z",
      "username": "pioluk",
    },
  },
  "type": "ADD_COMMENT_SUCCESS",
}
`;

exports[`comments actions should create an action representing a comment to remove 1`] = `
Object {
  "id": 1,
  "type": "REMOVE_COMMENT_SUCCESS",
}
`;

exports[`comments actions should create an action representing a request to remove a comment 1`] = `
Object {
  "id": 1,
  "type": "REMOVE_COMMENT",
}
`;

exports[`comments actions should create an action representing comment adding error 1`] = `
Object {
  "error": [Error: Comment adding failed],
  "type": "ADD_COMMENT_FAILURE",
}
`;

exports[`comments actions should create an action representing comment removal erorr 1`] = `
Object {
  "error": [Error: Comment removing failed],
  "type": "REMOVE_COMMENT_FAILURE",
}
`;

exports[`comments actions should create an action representing text to create comment 1`] = `
Object {
  "text": "Comment #1",
  "type": "ADD_COMMENT",
}
`;

exports[`comments actions should create an action to fetch comments 1`] = `
Object {
  "eventId": 1,
  "limit": 20,
  "offset": 0,
  "type": "FETCH_COMMENTS",
}
`;

exports[`comments actions should create an action to indicate comment fetching error 1`] = `
Object {
  "error": [Error: Error fetching comments],
  "type": "FETCH_COMMENTS_FAILURE",
}
`;

exports[`comments actions should create an action to send fetched comments 1`] = `
Object {
  "comments": Array [
    Object {
      "EventId": 1,
      "UserId": 1,
      "createdAt": "2016-12-03T12:00:00.000Z",
      "deletedAt": null,
      "id": 1,
      "text": "Lorem ipsum",
      "updatedAt": "2016-12-03T12:00:00.000Z",
      "user": Object {
        "createdAt": "2016-12-01T13:02:38.213Z",
        "email": "181423@edu.p.lodz.pl",
        "id": 1,
        "imageAvatar": null,
        "updatedAt": "2016-12-01T13:02:38.213Z",
        "username": "pioluk",
      },
    },
    Object {
      "EventId": 1,
      "UserId": 1,
      "createdAt": "2016-12-03T12:30:00.000Z",
      "deletedAt": null,
      "id": 2,
      "text": "Dolor Amet",
      "updatedAt": "2016-12-03T12:30:00.000Z",
      "user": Object {
        "createdAt": "2016-12-01T13:02:38.213Z",
        "email": "181423@edu.p.lodz.pl",
        "id": 1,
        "imageAvatar": null,
        "updatedAt": "2016-12-01T13:02:38.213Z",
        "username": "pioluk",
      },
    },
  ],
  "count": 2,
  "type": "FETCH_COMMENTS_SUCCESS",
}
`;
